### PR TITLE
Add gh mimetype database

### LINKED ISSUE #87: Přidat možnost vybrat databázi/knihovnu na mimetypes
Funkce `freeze` by měla brát nový argument, `guess_type`, což bude funkce, která se chová jako `mimetypes.guess_type` a použije se místo ní.

### PR DESCRIPTION
**Pull request by nappex**
fix #87

### CODE REVIEW

File: freezeyt/util.py [hunk 1]
```diff
@@ -146,3 +147,15 @@ def import_variable_from_module(
...
+def parse_mimetype_db(path):
+    """Parse mimetype: extesions dict structure from .json file,
+    which has a same structure as github pages mime-db.
+    """
+    with open(path) as file:
+        mime_db = json.load(file)
+
+    return {
+        k: v['extensions'] for k, v in mime_db.items() if v.get('extensions')
+    }
```
Comment (nappex) [inline freezeyt/util.py:161]: Another option could be
```suggestion
    return dict(
        filter(lambda t: t[1].get('extensions'), mime_db.items())
    )
```

File: tests/test_parse_mimetype_db.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+            "audio/webm": {
+                "source": "apache",
+                "compressible": False,
+                "extensions": ["weba"]
+            },
+        },
+        {
+            "audio/wav": ["wav"],
+            "audio/wave": ["wav"],
+            "audio/webm": ["weba"]
```
Comment (nappex) [inline tests/test_parse_mimetype_db.py:49]: Original mime_db for gh pages contains, separately same extensions for different mime-types

**Comment by nappex** [comment]
All extensions with more than one mimetype, occur in `db.json`

```python
bdoc {'application/bdoc', 'application/x-bdoc'}
mpp {'application/dash-patch+xml', 'application/vnd.ms-project'}
exe {'application/x-msdos-program', 'application/x-msdownload', 'application/octet-stream'}
dll {'application/x-msdownload', 'application/octet-stream'}
deb {'application/x-debian-package', 'application/octet-stream'}
dmg {'application/x-apple-diskimage', 'application/octet-stream'}
iso {'application/x-iso9660-image', 'application/octet-stream'}
msi {'application/x-msdownload', 'application/octet-stream'}
asc {'application/pgp-keys', 'application/pgp-signature'}
ac {'application/vnd.nokia.n-gage.ac+xml', 'application/pkix-attr-cert'}
rtf {'application/rtf', 'text/rtf'}
key {'application/x-iwork-keynote-sffkey', 'application/vnd.apple.keynote'}
numbers {'application/x-iwork-numbers-sffnumbers', 'application/vnd.apple.numbers'}
pages {'application/vnd.apple.pages', 'application/x-iwork-pages-sffpages'}
org {'application/vnd.lotus-organizer', 'text/x-org'}
stl {'application/vnd.ms-pki.stl', 'model/stl'}
pdb {'application/vnd.palm', 'application/x-pilot'}
rar {'application/vnd.rar', 'application/x-rar-compressed'}
prc {'application/x-mobipocket-ebook', 'model/prc', 'application/x-pilot'}
wmz {'application/x-msmetafile', 'application/x-ms-wmz'}
wmf {'application/x-msmetafile', 'image/wmf'}
emf {'application/x-msmetafile', 'image/emf'}
obj {'model/obj', 'application/x-tgif'}
xlf {'application/xliff+xml', 'application/x-xliff+xml'}
xml {'text/xml', 'application/xml'}
xsl {'application/xslt+xml', 'application/xml'}
3gpp {'video/3gpp', 'audio/3gpp'}
mp3 {'audio/mp3', 'audio/mpeg'}
m4a {'audio/mp4', 'audio/x-m4a'}
wav {'audio/wav', 'audio/x-wav', 'audio/wave'}
ra {'audio/x-pn-realaudio', 'audio/x-realaudio'}
bmp {'image/x-ms-bmp', 'image/bmp'}
jpm {'video/jpm', 'image/jpm'}
sub {'image/vnd.dvb.subtitle', 'text/vnd.dvb.subtitle'}
ico {'image/vnd.microsoft.icon', 'image/x-icon'}
pcx {'image/vnd.zbrush.pcx', 'image/x-pcx'}
x3db {'model/x3d+fastinfoset', 'model/x3d+binary'}
x3dv {'model/x3d+vrml', 'model/x3d-vrml'}
```

**Comment by encukou** [comment]
I guess you could first see GitHub Pages handles these: create a new repo, add files with these extensions to a `gh-pages` branch, turn on GH pages and look how GH serves them.

**Comment by nappex** [comment]
> I guess you could first see GitHub Pages handles these: create a new repo, add files with these extensions to a `gh-pages` branch, turn on GH pages and look how GH serves them.

I've created gh-pages site and result is saved [here](https://github.com/nappex/many_mimetypes_to_one_suffix/blob/main/gh_pages_responses)

There are also python scripts which I've created the result.

**Comment by nappex** [comment]
I am not sure how to continue because, we cannot ensure that gh-pages server will not decide to handle these content types by different way in future.

**Comment by nappex** [comment]
I am thinking what iterable will be the best option to use in case, where occurs more mimetypes to one suffix, I told on stream that I prefer `set` because of some probability of duplications, but there are not duplications for sure. `db.json` structure is that mimetypes are keys, so duplications are eliminated by this structure.
So I decided to use `set` by on time performance see:

```shell
python -m timeit -u usec -s 'a = set(range(1000, 1000000))' '100000 in a'
10000000 loops, best of 5: 0.0225 usec per loop
```

```shell
python -m timeit -s 'a = list(range(1000, 1000000))' '100000 in a'
500 loops, best of 5: 536 usec per loop
```

```shell
python -m timeit -s 'a = tuple(range(1000, 1000000))' '100000 in a'
500 loops, best of 5: 496 usec per loop
```
What is your opinion, what to use?

**Review by nappex — COMMENTED** [review]
During the pytest I have to help to myself to get information about `Wrongmimetypeerror`
```suggestion
        a = MultiError(self.failed_tasks.values())
        print(a.exceptions)
```

**Comment by nappex** [comment]
Hmm GH actions did not run. Strange!

**Comment by encukou** [comment]
I think that the performance differences between lists and sets won't matter at all if there won't be more than 5-10 elements.

File: freezeyt/freezer.py [hunk 3]
```diff
@@ -75,13 +88,14 @@ def check_mimetype(
     if url_path.endswith('/'):
         # Directories get saved as index.html
         url_path = 'index.html'
-    file_type = get_mimetype(url_path)
-    if not file_type:
-        file_type = default
+    file_types = get_mimetype(url_path)
+    if not file_types:
```
Comment (encukou) [inline freezeyt/freezer.py:92]:
```suggestion
    if file_types is None:
```

File: freezeyt/freezer.py [hunk 4]
```diff
@@ -75,13 +88,14 @@ def check_mimetype(
...
+        file_types = {default}
+
     headers = Headers(headers)
     mime_type, encoding = parse_options_header(headers.get('Content-Type'))
-    if file_type.lower() != mime_type.lower():
-        raise WrongMimetypeError(file_type, mime_type, url_path)
+    if mime_type.lower() not in file_types:
```
Comment (encukou) [inline freezeyt/freezer.py:97]:
```suggestion
    if mime_type.lower() not in (t.lower() for t in file_types):
```

File: freezeyt/freezer.py [hunk 5]
```diff
@@ -59,13 +60,25 @@ async def freeze_async(app, config):
...
 
+def github_mimetypes(suffixes_db):
 
+    def mimetypes(url: str) -> Optional[str]:
+        suffix = PurePosixPath(url).suffix
+        if suffix.startswith("."):
+            suffix = suffix[1:]
 
-def default_get_mimetype(url: str) -> Optional[str]:
+        return suffixes_db.get(suffix)
```
Comment (encukou) [inline freezeyt/freezer.py:70]:
```suggestion
        return suffixes_db.get(suffix.lower())
```

File: README.md [hunk 6]
```diff
@@ -314,7 +314,22 @@ The `get_mimetype`:
...
+There is also an option to load github database with informations, how to choose mimetype by the file extensions. Respectively, any database file in `json` format with same structure as github database can be used as source.
+
+If `json` database is used, then the default `get_mimetype` is switched to predefined github parser automatically.
+
+When it's configured `get_mimetypes` and `mimetype_db` suddenly, the configuration of `get_mimetypes` is omitted.
+
+Example of configuration `mimetype_db`:
+
+```yaml
+mimetype_db=path/to/db.json
```
Comment (encukou) [inline README.md:331]:
```suggestion
mime_db_file=path/to/db.json
```

File: freezeyt/freezer.py [hunk 7]
```diff
@@ -109,6 +126,20 @@ def parse_handlers(
     return result
 
 
+def parse_mime_db(mime_db: dict) -> Mapping:
```
Comment (encukou) [inline freezeyt/freezer.py:129]:
```suggestion
def convert_mime_db(mime_db: Mapping) -> dict[str, set[str]]:
```

File: freezeyt/util.py [hunk 8]
```diff
@@ -36,7 +36,7 @@ class WrongMimetypeError(ValueError):
     """MIME type does not match file extension"""
     def __init__(self, expected, got, url_path):
         super().__init__(
-            f"Content-type '{got}' is different from filetype '{expected}'"
+            f"Content-type '{got}' is different from allowed filetypes '{expected}'"
```
Comment (encukou) [inline freezeyt/util.py:39]:
```suggestion
            f"Content-type '{got}' is different from allowed MIME types '{expected}'"
```

File: tests/fixtures/app_wrong_mimetype/app.py [hunk 9]
```diff
@@ -24,3 +25,16 @@ def image():
     img_path = Path(__file__).parent / 'smile.png'
     img_bytes = img_path.read_bytes()
     return Response(img_bytes, mimetype='image/png')
+
+@app.route('/second_page.HTML')
```
Comment (encukou) [inline tests/fixtures/app_wrong_mimetype/app.py:29]:
```suggestion
@app.route('/capital_extension.HTML')
```

File: tests/test_mime_db_mimetype.py [hunk 10]
```diff
@@ -0,0 +1,184 @@
+import pytest
+import json
+
+from testutil import context_for_test
+from freezeyt import freeze
+from freezeyt.freezer import parse_mime_db, mime_db_mimetype
+
+
+PARSER_TEST_DATA = {
```
Comment (encukou) [inline tests/test_mime_db_mimetype.py:9]:
```suggestion
TEST_DATA = {
```

File: tests/test_mime_db_mimetype.py [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+            'mime_db_file': str(db_path)
+        }
+
+        freeze(module.app, freeze_config)
+
+    assert (builddir / 'index.html').exists()
+    assert (builddir / 'image.jpg').exists()
+
+
+def test_freeze_app_mock_parse_mime_db_uppercase(tmp_path, monkeypatch):
```
Comment (encukou) [inline tests/test_mime_db_mimetype.py:140]:
```suggestion
def test_freeze_app_mock_parse_mime_db_uppercase(tmp_path):
```

File: freezeyt/freezer.py [hunk 12]
```diff
@@ -62,29 +63,49 @@ async def freeze_async(app, config):
     '5xx': 'error',
 }
 
+def mime_db_mimetype(mime_db: dict, url: str) -> Optional[List[str]]:
+    """Determines file MIME type from file suffix. Decisions are made
+    by mime-db rules.
+    """
+    suffix = PurePosixPath(url).suffix
+    if suffix.startswith("."):
```
Comment (nappex) [inline freezeyt/freezer.py:71]: I am not sure if this condition is unnecessary

File: freezeyt/freezer.py [hunk 13]
```diff
@@ -197,12 +233,20 @@ def __init__(self, app, config):
...
             ('url_to_path', default_url_to_path)
         )
         for attr_name, default in CONFIG_DATA:
             setattr(self, attr_name, config.get(attr_name, default))
 
+        if self.mime_db_file:
+            with open(self.mime_db_file) as file:
+                mime_db = json.load(file)
+
+            mime_db = mime_db_conversion(mime_db)
```
Comment (nappex) [inline freezeyt/freezer.py:247]: I am not sure about same name, because content is same but structure is different

File: freezeyt/freezer.py [hunk 14]
```diff
@@ -62,29 +63,49 @@ async def freeze_async(app, config):
     '5xx': 'error',
 }
 
+def mime_db_mimetype(mime_db: dict, url: str) -> Optional[List[str]]:
+    """Determines file MIME type from file suffix. Decisions are made
+    by mime-db rules.
+    """
+    suffix = PurePosixPath(url).suffix
+    if suffix.startswith("."):
+        suffix = suffix[1:]
```
Comment (jiri-one) [inline freezeyt/freezer.py:72]: [pathlib.PurePath.suffix](https://docs.python.org/3/library/pathlib.html#pathlib.PurePath.suffix) the return value from suffix property is always (if suffix is found) with dot, so the code should be only `suffix = PurePosixPath(url).suffix[1:]` and if statement is unnecessary

Comment (nappex) [reply freezeyt/freezer.py:72]: Thanks for feedback, write me your email registered in github and I will add as coauthor

Comment (jiri-one) [reply freezeyt/freezer.py]: NP @nappex, I have written my mail on Slack to PM ;). Have a nice day, Jiří

File: freezeyt/freezer.py [hunk 15]
```diff
@@ -62,29 +63,51 @@ async def freeze_async(app, config):
...
-    if file_type.lower() != mime_type.lower():
-        raise WrongMimetypeError(file_type, mime_type, url_path)
+    headers_mimetype, encoding = parse_options_header(
+        headers.get('Content-Type')
+    )
+
+    if isinstance(file_mimetypes, str):
+        raise TypeError(
+            f"get_mimetype result must not be a {type(file_mimetypes)!r}"
+        )
```
Comment (nappex) [inline freezeyt/freezer.py:108]: Is good idea to use
```python
{type(file_mimetypes)!r}
```
or just
```python
string
```
is better?

Comment (encukou) [reply freezeyt/freezer.py:108]: I think just `string` is best

File: freezeyt/freezer.py [hunk 16]
```diff
@@ -62,29 +63,50 @@ async def freeze_async(app, config):
...
-        raise WrongMimetypeError(file_type, mime_type, url_path)
+    headers_mimetype, encoding = parse_options_header(
+        headers.get('Content-Type')
+    )
+
+    if isinstance(file_mimetypes, str):
+        raise TypeError("get_mimetype result must not be a string")
+
+    if headers_mimetype.lower() not in (m.lower() for m in file_mimetypes):
+        raise WrongMimetypeError(file_mimetypes, headers_mimetype, url_path)
```
Comment (Glutexo) [inline freezeyt/freezer.py:109]:
```suggestion
        if isinstance(file_mimetypes, str):
            # This looks like the place for a nicer error message.
        raise WrongMimetypeError(file_mimetypes, headers_mimetype, url_path)
```

File: freezeyt/freezer.py [hunk 17]
```diff
@@ -62,29 +63,50 @@ async def freeze_async(app, config):
...
+        return [file_mimetype]
 
 
 def check_mimetype(
     url_path, headers,
-    default='application/octet-stream', *, get_mimetype=default_get_mimetype,
+    default='application/octet-stream', *, get_mimetype=default_mimetype,
 ):
+    """Compare mimetype sent from headers with file mimetype guessed
+    from its suffix.
```
Comment (encukou) [inline freezeyt/freezer.py:91]:
```suggestion
    """Ensure mimetype sent from headers with file mimetype guessed
    from its suffix.
    Raise WrongMimetypeError if they don't match.
```

**Review by encukou — APPROVED** [review]

**Pull request merged**